crates/ra_analysis/src/completion/completion_context.rs

```diff
@@ -145,6 +152,23 @@ impl<'a> CompletionContext<'a> {
                 }
             }
         }
+        if let Some(_field_expr) = ast::FieldExpr::cast(parent) {
+            self.dot_receiver = self
+                .leaf
+                .ancestors()
+                .take_while(|it| it.kind() != SOURCE_FILE && it.kind() != MODULE)
+                .find_map(ast::FieldExpr::cast)
+                .and_then(ast::FieldExpr::expr);
```

Comment:
I feel that this could break, because leaf can be an arbitrary thing:

```
foo
  /* comment */
  .
  /* more comments */
  cursor_here
```

We need to figure out a more robust pattern here. In intelliJ, this  "patched file"

```
        let file = {
            let edit = AtomTextEdit::insert(offset, "intellijRulezz".to_string());
            original_file.reparse(&edit)
        };
```

is a long-lived thing and all completion infra uses it. However, this leads to a ton of bugs when you use patched file where the original file is expected. That's why in ra I try to strictly separate the "gather context" and "compute completion" bits.


I think what would work here is to find the receiver expr in the "patched" file, get its range, and then find an expression with this range in the original file.

Reply:
Ok, done :white_check_mark: